zerver/views/message_summary.py

```diff
@@ -0,0 +1,131 @@
+import json
+from typing import Any
+
+import litellm
+from django.conf import settings
+from django.http import HttpRequest, HttpResponse
+from django.utils.timezone import now as timezone_now
+from pydantic import Json
+
+from analytics.lib.counts import COUNT_STATS, do_increment_logging_stat
+from zerver.lib.narrow import NarrowParameter
+from zerver.lib.response import json_success
+from zerver.lib.typed_endpoint import typed_endpoint
+from zerver.models import UserProfile
+from zerver.views.message_fetch import get_messages_backend
+
+# Maximum number of messages that can be summarized in a single request.
+MAX_MESSAGES_SUMMARIZED = 100
+OUTPUT_COST_PER_TOKEN = 5
+INPUT_COST_PER_TOKEN = 1
+
+
+def format_conversation(zulip_messages: list[dict[str, Any]]) -> str:
+    # Note: Including timestamps seems to have no impact; including reactions
+    # makes the results worse.
+
+    zulip_messages_list = [
+        {"sender": message["sender_full_name"], "content": message["content"]}
+        for message in zulip_messages
+    ]
+    return json.dumps(zulip_messages_list)
+
+
+def make_message(content: str, role: str = "user") -> dict[str, str]:
+    return {"content": content, "role": role}
+
+
+def get_max_summary_length(conversation_length: int) -> int:
+    return min(6, 4 + int((conversation_length - 10) / 10))
+
+
+@typed_endpoint
+def get_messages_summary(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    *,
+    narrow: Json[list[NarrowParameter] | None] = None,
+) -> HttpResponse:
+    if not user_profile.is_realm_admin:
+        return json_success(request, {"summary": "Feature limited to admin users for now."})
+
+    # Since there will always be a limit to how much data we want the LLM to process
+    # at once, due to API limits or performance reasons at least, we generate summaries
+    # for the messages in chunks. We will be using Rolling Summaries for this purpose.
+    # Rolling Summaries are summaries that are generated for a fixed number of messages
+    # at a time, and then the next summary is generated for the next fixed number of
+    # messages with the previous summary as the starting point. This way, we can
+    # generate summaries for new messages in a single pass.
+    # TODO: Come up with a plan to store these summaries in the database.
+
+    messages_response = get_messages_backend(
+        request,
+        user_profile,
+        narrow=narrow,
+        anchor_val="newest",
+        num_before=MAX_MESSAGES_SUMMARIZED,
+        client_gravatar=False,
+        apply_markdown=False,
+    )
+    zulip_messages = json.loads(messages_response.content).get("messages", [])
```

Comment:
Hmm, I'd have thought we'd do this at a lower layer, like `fetch_messages`, and avoid doing a bunch of extra work that's required for this use case.

Reply:
`get_messages_backend` has some parameter check and updates for `narrow` which I don't think we want to skip here. 
What do you think about extracting the complete `get_messages_backend` into a separate funtion and adding a parameter to it which will skip any extra steps which we don't want here?

Reply:
I'd rather just extract the `narrow` parameter checks into a function called by both `get_messages_backend` and here. It's a lot easier to read code that just doesn't call a chunk of work we don't need, then to read something with a parameter to skip a bunch of stuff.

And `GET /messages` is one of our most important code paths, so keeping it nice is important.

Reply:
Done.